### ISSUE TITLE
Which scipy.stats functions can be used with xarray.reduce()?

### ISSUE THREAD

**Issue by atreguier** [pydata/xarray#6861]
### What is your issue?

I would like to use scipy.stats.skew and scipy.stats.kurt with multidimensional xarrays, and also with results of the "coarsen" function. 

I don't see information about which functions can be used in the documentation, and I find that the errors that I get are difficult to understand (for me).

I'm not sure it is a bug, though, so I report it as an issue.  Here is a test code: 

import xarray as xr
import scipy.stats
import numpy as np
#  define random dataarrays, 1D and 2D -- 
da = xr.DataArray( np.random.rand(100),dims=['x'])
da2 = xr.DataArray( np.random.rand(10,10),dims=['x','y'])
#  Work on the whole one dimensional array, OK -- 
print('skewness 1D:',da.reduce(func=scipy.stats.skew).values)
#
#  with 2D array it does not work -- 
print('skewness 2D:',da2.reduce(func=scipy.stats.skew).values)
# This gives error: ValueError: dimensions () must have the same length as the number of data dimensions, ndim=1 --
#
#  try to compute the skewness in bins of 10 using coarsen: does not work --
daskew=da.coarsen(x=10).reduce(func=scipy.stats.skew)
#  gives error: TypeError: tuple indices must be integers or slices, not tuple -- 
#
daskew=da.coarsen(x=10).reduce(func=scipy.stats.skew,axis=None)
#  gives error:  TypeError: skew() got multiple values for keyword argument 'axis'

**Comment by keewis** [comment]
I agree The general requirement on reduce functions is that they behave like `numpy` ufuncs, e.g. `np.mean` (`xarray` does not support the other ways to call ufuncs, e.g. `np.mul.reduceat`). As far as I know, that means that they take both the data and a `axis` argument that can either be a int, a tuple of int, or `None` (the default).

The `scipy.stats` functions don't seem to be true ufuncs: unlike `numpy`'s functions, they will by default not reduce along all axes (`axis=None`) but along the first axis (`axis=0`).

To be able to use them with the standard reduce, you'd have to write a wrapper that changes that default:
```python
def skew(data, *, axis=None, **kwargs):
    return scipy.stats.skew(data, axis=axis, **kwargs)

da2.reduce(skew)  # works
```

They also don't know what to do with a tuple of int as axis. Since `coarsen(...).reduce(...)` uses that, it fails:
```python
np.mean(da2.data, axis=(0, 1))
scipy.stats.skew(da2.data, axis=(0, 1))  # fails
```
I didn't try to investigate further (so I suspect this can be simplified quite a bit), but we can modify our wrapper to transform the data into a form that `scipy.stats.skew` can understand:
```python
def stack_axes(arr, axis):
    if isinstance(axis, int):
        axis = (axis,)

    remainder = tuple(a for a in range(arr.ndim) if a not in axis)
    transposed = np.transpose(arr, remainder + axis)
    new_shape = arr.shape[: len(remainder)] + (-1,)
    return np.reshape(transposed, new_shape)

def skew(data, *, axis=None, **kwargs):
    if isinstance(axis, tuple):
        data = stack_axes(data, axis=axis)
        axis = -1

    return scipy.stats.skew(data, axis=axis, **kwargs)

da.coarsen(x=10).reduce(func=skew)  # works
da.reduce(func=skew)  # still works
```

Finally, the last call fails because `coarsen(...).reduce` already passes a `axis` kwarg so passing one again will make python complain.

**Comment by atreguier** [comment]
Thanks, this is very helpful!

**Comment by OriolAbril** [comment]
In case it is useful, we created https://einstats.python.arviz.org/en/latest/ a while back, which we use in [ArviZ](https://python.arviz.org/en/stable/) but thought could be useful outside Bayesian modeling too so we decided to make into its own small library. It defines xarray-friendly wrappers for some linalg, stats and einops operations. It wouldn't help with coarsen as the functions don't take `axis` arguments, but they do work with GroupBy objects.

For now it mostly has things we use but happy to include more, and [skew](https://einstats.python.arviz.org/en/latest/api/generated/xarray_einstats.stats.skew.html) and [kurtosis](https://einstats.python.arviz.org/en/latest/api/generated/xarray_einstats.stats.kurtosis.html) are already there.
